app/api/companies.py

```diff
@@ -138,43 +142,89 @@ def get_company(id):
-
-    response = jsonify(company.to_dict())
+    try:
+        result = CompaniesValidationSchema().load(data)
+    except ValidationError as err:
+        print(err.messages)
+        print(err.valid_data)
+        return bad_request(err.messages)
+
+    # TODO: Validate if company name is already in DB?
+
+    new_company = Companies(company_name=result['company_name'], logo_image_src=result['logo_image_src'],
+                            website=result['website'], year=result['year'], company_size=result['company_size'])
+
+    # Check if city is already in DB:
+    city = Cities.query.filter_by(
+        city_name=result['city_name'].title()).first()
+    if city is not None:
+        new_company.city_id = city.city_id
+        db.session.add(new_company)
+        db.session.commit()
+    else:
+        new_city = Cities(
+            city_name=result['city_name'].title(), region=result['region'])
+        new_city.company.append(new_company)
+        db.session.add(new_city)
+        db.session.commit()
+
+    # Insert Meta Data:
+    insert_meta(result['disciplines'], 'disciplines', new_company.company_id)
+    insert_meta(result['branches'], 'branches', new_company.company_id)
+    insert_meta(result['tags'], 'tags', new_company.company_id)
+
+    response = jsonify(new_company.to_dict())
     response.status_code = 201
     response.headers['Location'] = url_for(
-        'api.get_company', id=company.company_id)
+        'api.get_company', id=new_company.company_id)
 
     cache.clear()
     return response
 
 
 # TODO: Create a specific error message when selecting a non-existing company_id
-@bp.route('/v1/companies/<int:id>', methods=['PUT'])
+@bp.route('/v1/companies/<int:id>', methods=['PATCH'])
 @jwt_required()
 def update_company(id):
-    company = Companies.query.get_or_404(id)
+
     data = request.get_json() or {}
-    # TODO: Validation
-    if False:
-        return bad_request('false')
-    # TODO: Finalize the from_dict method
-    company.from_dict(data, new_company=False)
-    db.session.commit()
+
+    try:
+        result = CompaniesPatchSchema().load(data, partial=True)
+    except ValidationError as err:
+        return bad_request(err.messages)
+
+    exceptions = ['city_name', 'disciplines', 'branches', 'tags']
+    for field in exceptions:
+        if field in data:
+            if field == 'city_name':
+                city_dict = insert_city(data)
+                data['city_id'] = city_dict['city_id']
+                data.pop('city_name')
+                data.pop('region')
+            if field in ['disciplines', 'branches', 'tags']:
+                # TODO: Only removes the records from the companies_meta table, not the actual records in the Meta table (as these could still be in use by other companies)
+                db.session.execute("DELETE FROM companies_meta WHERE companies_meta.company_id = :id AND companies_meta.meta_id IN (SELECT Meta.meta_id FROM Meta WHERE Meta.type = :type)", {
```

Comment:
Yeah - this works for now

I suppose we you could conditionally delete the Meta record if it is no longer used anymore 
You would have to make a subquery on the companies_meta table to check this. 

Since the requirements regarding what we want to do with the meta is not so clear at the moment we can just go with this for now 👍

Reply:
Indeed, something worth to look into at a later stage. Will keep the comment there for now to keep reminding me of this issue.